### ISSUE TITLE
Issue while mapping null to a property.

### ISSUE THREAD

**Issue by Alysaar** [agileobjects/AgileMapper#176]
Hello, I have an issue when attempting to map a null object to a property.
Here is an example :

```csharp
class Program
{
    static void Main(string[] args)
    {
        var mapper = Mapper.CreateNew();
        mapper.WhenMapping.ThrowIfAnyMappingPlanIsIncomplete(); // needed in my codebase

        mapper.WhenMapping.From<SourceClass>().To<DestinationClass>()
            .Map((source, _) => GetSomeProperty(source)).To(dest => dest.TheProperty);

        var result = mapper.Map(new SourceClass()).ToANew<DestinationClass>();
    }

    private static DestinationPropertyClass GetSomeProperty(SourceClass source)
    {
        // some operation that uses information from SourceClass to find / build a DestinationPropertyClass, and can return null
        return null;
    }

    public class SourceClass { }

    public class DestinationClass
    {
        public DestinationPropertyClass TheProperty { get; set; }
    }

    public class DestinationPropertyClass
    {
        public int SomeProperty { get; set; }
    }
}
```

I was expecting to get an instance of `DestinationClass` with its `TheProperty` set to null.
Instead I had a NullReferenceException thrown when attempting to apply the mapping.

Is this a bug or am I doing something wrong ?

Thank you.

**Comment by SteveWilkes** [comment]
Hi!

Thanks for reporting this - it's a bug.

We might actually need some extra configuration to make the intent clear - let's say you're configuring from

```csharp
SourceClass
  + TheProperty: Nested

Nested
  + Value: int
```

...to:

```csharp
DestinationClass
  + TheProperty: Nested
```

If you configured:

```csharp
mapper.WhenMapping
    .From<SourceClass>().To<DestinationClass>()
    .Map((source, _) => GetSomeProperty(source)).To(dest => dest.TheProperty);
```

...would you expect the mapper to map `SourceClass.TheProperty.Value` to `DestinationClass.TheProperty.Value` _after_ calling `GetSomeProperty()`, or would you expect it to stop mapping `TheProperty` at that point, assuming `GetSomeProperty()` had mapped everything required?

Would this (or something like this) make it clearer, d'you think?

```csharp
mapper.WhenMapping
    .From<SourceClass>().To<DestinationClass>()
    .Map((source, _) => GetSomeProperty(source)).To(dest => dest.TheProperty)
    .And.EndThatMapping();
```

Cheers,

Steve

**Comment by Alysaar** [comment]
Hi, thank you for responding.

I'm going to answer from the point of view of my current need in my codebase, so it might not be what is best for all cases.

> ...would you expect the mapper to map `SourceClass.TheProperty.Value` to `DestinationClass.TheProperty.Value` after calling `GetSomeProperty()`

Yes that is exactly what I expected when using this.

> or would you expect it to stop mapping `TheProperty` at that point, assuming `GetSomeProperty()` had mapped everything required

No, I would not assume that. If the `GetSomeProperty` method has failed to map one or more properties of `TheProperty`, I would preffer getting an error (since I'm using the `ThrowIfAnyMappingPlanIsIncomplete()` option). 
Also In my example the `GetSomeProperty` method returns an instance of `DestinationPropertyClass` but in actuality in my codebase it returns an instance of a different class that contains some properties that would then be mapped to the `DestinationClass.TheProperty`'s properties (and once again I would prefer an error if some properties are missing).

> Would this (or something like this) make it clearer, d'you think?
> ```csharp
> mapper.WhenMapping
>     .From<SourceClass>().To<DestinationClass>()
>     .Map((source, _) => GetSomeProperty(source)).To(dest => dest.TheProperty)
>     .And.EndThatMapping();
> ```

I'm not sure I understand what you are going for here. Is the goal to ignore errors caused by  `ThrowIfAnyMappingPlanIsIncomplete()` for just this property ? If so, wouldn't it be simpler to just `Ignore` this property and then do a second mapping for it using `OnTo` or something ?

What do you think ?

**Comment by SteveWilkes** [comment]
That probably answers the question as to whether it's clearer like that, then :P

The default behaviour would (/ should) be to call `GetSomeProperty` to get the inital `Nested` value, then map the `Nested.Value` property. I wondered if it was clear that was what it was doing, and if an option to tell the mapper that `GetSomeProperty` performs the entire `Nested` mapping - so not to map `Nested.Value` - would be useful.

Sounds like your instinct on what it should do is the same as mine, so I won't add that option yet. Cheers!

**Comment by bslbckr** [comment]
> Would this (or something like this) make it clearer, d'you think?
> 
> ```cs
> mapper.WhenMapping
>     .From<SourceClass>().To<DestinationClass>()
>     .Map((source, _) => GetSomeProperty(source)).To(dest => dest.TheProperty)
>     .And.EndThatMapping();
> ```
> 
> Cheers,

I would have loved the `And.EndThatMapping`-construct also after creating instances with the `CreateInstances-Using` factory method.

**Comment by SteveWilkes** [comment]
> I would have loved the `And.EndThatMapping`-construct also after creating instances with the `CreateInstances-Using` factory method.

Yeah, I'd considered using it there, too :) As an alternative, was thinking I could add:

```csharp
mapper.WhenMapping
    .From<SourceClass>().To<DestinationClass>()
    .MapInstancesOf<Nested>().Using(ctx => CreateNested(ctx));
```

**Comment by bslbckr** [comment]
Yes, this would also add to the clarification and verbosity of the api.

**Comment by SteveWilkes** [comment]
The original issue - unhandled null values from configured data source methods - is now fixed in the latest v1.6 branch code. Release to follow! Thanks for the feedback!

**Comment by SteveWilkes** [comment]
The fix for unhandled null values from configured data source methods is included in v1.6, which is now available [on NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper/1.6.0).

Thanks again for the feedback!

**Comment by SteveWilkes** [comment]
`MapInstancesUsing` and `MapInstancesOf<T>` are both supported in the latest v1.7 branch code - preview release to follow.

```
mapper.WhenMapping
    .From<SourceClass>().To<DestinationClass>()
    .MapInstancesOf<Nested>().Using(ctx => CreateNested(ctx));
```

**Comment by SteveWilkes** [comment]
`MapInstancesUsing` and `MapInstancesOf<T>` are included in v1.7-preview1, which is now available [on NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper/1.7.0-preview1).
